### PR TITLE
Boutiques

### PR DESCRIPTION
**Pull request by arthursw**
### New frontend with Angular-cli 8.3.19

The front end is both accessible for development mode (at `https://shanoir-ng-nginx/dev/`) and production mode (at `https://shanoir-ng-nginx/shanoir-ng/`).

#### Front-dev container
 
The docker container front-dev is just a node image with angular cli:

front-dev/Dockerfile:

```
FROM node:12.10.0-stretch

WORKDIR /app

RUN npm install -g @angular/cli
```

A docker volume links `shanoir-ng-front` to the containers `app/` directory.
When launched, this container runs `ng serve` to serve the development version of the front end.

#### Nginx configuration

Nginx serves production files statically at `/shanoir-ng/` and proxies the development version to `http://front-dev:4200/`.

`nginx.conf`:
```
[...]
		location /shanoir-ng/ {
			alias   /etc/nginx/html/;
			index  index.html;
			try_files $uri$args $uri$args/ $uri $uri/ /index.html =404;
		}
		# Front development with angular-cli server
		location /dev/ {
			proxy_pass 				http://front-dev:4200/;

			proxy_http_version 1.1;
			proxy_set_header Upgrade $http_upgrade;
			proxy_set_header Connection "Upgrade";
		}
[...]
```

#### Compile front end for production

Use `ng build --prod` to build the front end for production. This will generate production static files in the `dist/` directory (without removing and replacing it, thanks to the `"deleteOutputPath": false,` build option in `angular.json`). Since a volume links `shanoir-ng-front/dist/` to the nginx container `/etc/nginx/html/`, the production front end will be updated.

Note: `src/index.html` is used for the development version, and `src/index.prod.html` for production. The only different is the base href:
in `src/index.html` (development mode): 
```
  <base href="/dev/">
```
in `src/index.prod.html` (production mode):
```
  <base href="/shanoir-ng/">
```

#### Angular AOT compilation

Here is a subset of the list of bugs I fixed because Ahead-Of-Time builds forbids access to private attributes in .html (see [AOT metadata errors](https://angular.io/guide/aot-metadata-errors) ): [Bugs AOT Build.txt](https://github.com/fli-iam/shanoir-ng/files/3859606/Bugs.AOT.Build.txt)

#### Install notes

You can safely remove `node_modules` and run `npm install` to have a clean installation.

### CODE REVIEW

**Comment by arthursw** [comment]
I just realized that maven does not compile any more since I removed the unused `shanoir-ng-front/pom.xml`. You can just remove `shanoir-ng-front` from the `shanoir-ng-parent/pom.xml` and it should work.

**Comment by Nunien** [comment]
@arthursw Arthur, until the pull request is merged you can update it by simply pushing new commits into your local branch (that the pull request is made from)

File: docker-compose/nginx/Dockerfile [hunk 1]
```diff
@@ -14,4 +14,4 @@ FROM nginx
 
 COPY nginx.conf /etc/nginx/nginx.conf
 
-COPY webapp/ /etc/nginx/html/
\ No newline at end of file
+# COPY webapp/ /etc/nginx/html/
```
Comment (jcomedouteau) [inline docker-compose/nginx/Dockerfile]: Isn't this for development only ?

File: shanoir-ng-front/src/app/boutiques/tool.service.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import { ToolInfo } from './tool.model';
+import { EntityService } from '../shared/components/entity/entity.abstract.service';
+import { ServiceLocator } from '../utils/locator.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ToolService extends EntityService<ToolInfo> {
+
+  API_URL = 'http://localhost:9000/tool';
```
Comment (jcomedouteau) [inline shanoir-ng-front/src/app/boutiques/tool.service.ts]: Is it always a localhost service ?

File: shanoir-ng-front/src/app/boutiques/invocation-gui/parameter-group/parameter-group.component.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  @Input() parameterGroup: ParameterGroup;
+  @Input() formGroup: FormGroup;
+  @ViewChildren(ParameterComponent) parameterComponents !: QueryList<ParameterComponent>;
+
+  selectedParameterId: string = null;
+
+  constructor() { }
+
+  ngOnInit() {
```
Comment (jcomedouteau) [inline shanoir-ng-front/src/app/boutiques/invocation-gui/parameter-group/parameter-group.component.ts]: le "implements OnInit" est-il utile du coup ?

**Comment by michaelkain** [comment]
Hi @julien-louis, it would be great to analyze the contribution inside this PR on angular-cli please.
I will check for the backend, what it implies.

**Comment by jcomedouteau** [comment]
Won't be done.

**Pull request closed without merging**